### PR TITLE
Validate resource requests for zero or negative values

### LINKED ISSUE #1390: Resource values (cpu, memory, gpu) should be validated
**Describe the issue**
When editing a node's properties, I am able to enter negative values for CPU and GPU. Note that I can enter a negative value for memory, but it is caught _on the server_.  Also, I can enter invalid values (32 GPU) when there aren't any, and the pipeline's execution will succeed w/o error.

**To Reproduce**
1. Open a pipeline's node properties and enter -8 for the CPU.  Although there is a warning, the value is permitted.  Also enter a value of some large number of GPUs (e.g., 37)
2. Execute the pipeline successfully.

I tried this on KFP and local.  Did not try on Airflow, but I suspect the same issue.

**Expected behavior**
At a minimum, invalid values like negative numbers should be caught on the client.  I understand it might be harder (or impossible) to validate actual numbers of CPUs and GPUs since a runtime may be involved.

### PR DESCRIPTION
**Pull request by akchinSTC**
Fixes #1390

 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

File: elyra/pipeline/pipeline.py [hunk 1]
```diff
@@ -142,6 +142,13 @@ def env_vars_as_dict(self, logger: Optional[object] = None) -> Dict:
                         print(f"Could not process environment variable entry `{nv}`, skipping...")
         return envs
 
+    def validate_resource_value(self, name: str, value):
+        if value is None or int(value) > 0:
```
Comment (lresende) [inline elyra/pipeline/pipeline.py:146]:
```suggestion
        if value and int(value) > 0:
```

**Comment by bourdakos1** [comment]
It should not let you run the pipeline unless you have positive resources now

File: packages/pipeline-editor/src/PipelineEditorWidget.tsx [hunk 2]
```diff
@@ -1172,6 +1172,15 @@ export class PipelineEditor extends React.Component<
     ) {
       validationErrors.push('no runtime image');
     }
+    if (node.app_data.cpu <= 0) {
+      validationErrors.push('CPU must be greater than 0');
+    }
+    if (node.app_data.gpu <= 0) {
```
Comment (kevin-bates) [inline packages/pipeline-editor/src/PipelineEditorWidget.tsx:1178]: Seems like 0 GPU should be allowed.  The same might go for CPU, although I don't really know if something that uses GPUs can exclude the use of CPUs (I kinda doubt it).

Comment (bourdakos1) [reply packages/pipeline-editor/src/PipelineEditorWidget.tsx:1178]: oh yea that makes sense, I was just going off of what the ui controls were validating for, I'll update those too

so, like this?
cpu > 0
memory > 0 
gpu >= 0

Comment (kevin-bates) [reply packages/pipeline-editor/src/PipelineEditorWidget.tsx:1178]: Those make sense to me - thanks.

File: elyra/pipeline/pipeline.py [hunk 3]
```diff
@@ -70,9 +70,9 @@ def __init__(self, id, type, classifier, filename, runtime_image, memory=None, c
         self._inputs = inputs or []
         self._outputs = outputs or []
         self._parent_operations = parent_operations or []
-        self._cpu = cpu
-        self._gpu = gpu
-        self._memory = memory
+        self._cpu = self.validate_resource_value("CPU", cpu)
+        self._gpu = self.validate_resource_value("GPU", gpu)
```
Comment (kevin-bates) [inline elyra/pipeline/pipeline.py:74]: Regarding my earlier comment relative to the front-end, it seems like we should tolerate a zero value for GPU.  Sorry, this complicates the check and messaging.

File: conftest.py [hunk 4]
```diff
@@ -15,5 +15,5 @@
 #
 import pytest
 
-pytest_plugins = ["jupyter_server.pytest_plugin"]
+#pytest_plugins = ["jupyter_server.pytest_plugin"]
```
Comment (kevin-bates) [inline conftest.py:18]: Is this causing issues?  I _think_ it's necessary for tests that use fixtures from jupyter_server.  The metadata tests use a number of these.  That said, I think I've found this to be unnecessary, but that never made sense to me - so I'm curious if you found some adverse to defining this.  (The pytest fixture stuff is still a bit of magic to me.)

Comment (akchinSTC) [reply conftest.py:18]: My mistake, I left this in there. I've had to comment this out when testing locally due to an error being thrown when using pytest. Let me re-enable it. 
I'm kind of in the same boat you are, this file has been unnecessary for me but needed for others.

File: packages/pipeline-editor/src/properties.json [hunk 5]
```diff
@@ -298,7 +298,7 @@
               "condition": {
                 "parameter_ref": "gpu",
                 "op": "greaterThan",
-                "value": 0
+                "value": -1
```
Comment (kevin-bates) [inline packages/pipeline-editor/src/properties.json:301]: The `validation.fail_message.message.default` field above should read to the effect of "GPU must be greater than or equal to 0."

Comment (bourdakos1) [reply packages/pipeline-editor/src/properties.json:301]: oh yep, good catch

**Review by kevin-bates — APPROVED** [review]
Thanks @akchinSTC and @bourdakos1!

**Review by lresende — APPROVED** [review]

**Pull request merged**